File: Python3/Combination Sum.py
# 104ms, 74.87%
class Solution:
    def combinationSum(self, candidates, target):
        candidates.sort()
        res, com = [], []
        self.calculate(candidates, target, res, com, 0)
        return res
    
    def calculate(self, candidates, target, res, com, begin):
        if not target:
            res.append(com[:])
            return
        for i in range(begin, len(candidates)):
            if candidates[i] > target: break
            com.append(candidates[i])
            self.calculate(candidates, target - candidates[i], res, com, i)
            com.pop()
    

File: Java/Combination Sum.java
// 3ms, 81.92%; 39.5MB, 82.57%
class Solution {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new LinkedList<>();
        LinkedList<Integer> tmp = new LinkedList<>();
        dfs(candidates, target, res, tmp, 0);
        return res;
    }

    private static void dfs(int[] candidates, int target, List<List<Integer>> res, LinkedList<Integer> tmp, int index) {
        while (index < candidates.length && candidates[index] <= target) {
            tmp.offerLast(candidates[index]);
            if (candidates[index] == target) {
                res.add(List.copyOf(tmp));
                tmp.pollLast();
                break;
            } else {
                dfs(candidates, target - candidates[index], res, tmp, index++);
            }
            tmp.pollLast();
        }
    }
}
